add function to create breeding event

HenrikeW opened the pull request:
When clicking on "Skapa", 
- a breeding event is created,
- it's updated with birth information and
- the individual is created.
Three API-requests chained.

The newly created individual will have the field "breeding" which connects it with information about its parents and litter size.

HenrikeW committed:
add function to create breeding event

pontus commented on frontend/src/individual_add.tsx:
```diff
@@ -158,6 +159,51 @@ export function IndividualAdd({
+      };
+
+      const breedingEvent = await post("/api/breeding", breeding);
+
```
Error handling? 

And we should probably check if this breeding exist before trying to create it (as mentioned, I don't know if the API is there, but we can add if needed).

HenrikeW replied:
Right. My thought was to have this one, pretty generic error handling for both requests:
```
 } else {
        userMessage(
          // register_birth in data_access.py requires a litter size.
          // Thus, the field must be mandatory in the form. Okay for föreningen?
          "Något gick fel. Kolla så att du har lagt till mor, far, kullstorlek och födelsedatum.",
          "error"
        );
      }
```
It will be triggered if something in one of the requests goes wrong, but maybe it's not so beautiful. I can add more specific error handling to the first request to prevent sending a second request that is meant to fail from the very beginning. 

Also, the backend checks if there already is a breeding, it looks like this in `register_breeding` in `data_access.py`: 
```
exists = (
        Breeding.select()
        .where(Breeding.father == father)
        .where(Breeding.mother == mother)
        .where(Breeding.breed_date == breed_date)
        .count()
    )
    if exists:
        return {"status": "error", "message": "Breeding already registered"}
```
But I would need the endpoint to send the breeding's id even in this case. For now, it's only send in case a new breeding is created. So that's something that needs to be added, and I don't think I can do that myself.

pontus replied:
I don't see a problem with a generic error message, but if I understand things correctly, that relies on doing the second request and failing - is that incorrect?

(I would possibly expect to get some other error because of `breeding_id` not being present in `breedingEvent`, but I guess that doesn't happen.)

pontus replied:
Regarding backend; if it simplifies I guess we can change the add breeding api endpoint to allow existing breeding to just return that. The alternative would be to get the breeding for the herd and check if it exists already.

HenrikeW replied:
So now we check if the breeding exists by getting all the breedings for the current herd. Only there is no breeding a request is sent to create one.

HenrikeW committed:
adding error handling to breeding and birth request

more specific handling of different errors.
for /api/breeding still something missing in the backend.

pontus commented on frontend/src/individual_add.tsx:
```diff
@@ -158,6 +159,106 @@ export function IndividualAdd({
+        );
+      } else if (
+        breedingEvent.status == "error" &&
+        breedingEvent.massage == "Forbidden"
```
```suggestion
        breedingEvent.message == "Forbidden"
```

pontus replied:
But I'm sorry if that was unclear, I don't have a problem with generic messages. And if we want to offer better messages in the frontend, I think we should define error codes rather than rely on strings like htis.

HenrikeW replied:
Yes, I agree! I'd also love to have some error codes instead of strings. It would be easier and also more in line with standards. Now even an error from the backend becomes a "200 OK" in the browser.

pontus commented on frontend/src/individual_add.tsx:
```diff
@@ -158,6 +159,106 @@ export function IndividualAdd({
+          "error"
+        );
+      }
+
```
It will still (currently) continue here, right?

HenrikeW committed:
get all breedings for origin herd

only attempt to create a new one if no breeding exists yet.
also moving the create breeding part to its own function.
error handling not completely tested yet.

HenrikeW committed:
improve error handling

pontus commented on frontend/src/individual_add.tsx:
```diff
@@ -187,59 +200,97 @@ export function IndividualAdd({
+          const breedingMatch = breedings.breedings.find((item) => {
+            item.mother == currentBreeding.mother &&
+              item.father == currentBreeding.father &&
+              item.breed_date == currentBreeding.date;
```
```suggestion
              (item.breed_date == currentBreeding.date ||
               item.birth_date == individual.birth_date);
```

?

HenrikeW replied:
Good catch!

pontus commented on frontend/src/individual_add.tsx:
```diff
@@ -187,59 +200,97 @@ export function IndividualAdd({
+              birthEvent.message == "Not logged in"
+            ) {
+              userMessage(
+                "Du är inte inloggad. Logga in och försök igen.",
+                "warning"
+              );
+            } else if (
+              birthEvent.status == "error" &&
+              birthEvent.message.includes("litter size" || "Litter size")
+            ) {
+              userMessage("Ange en kullstorlek större än null.", "warning");
+            } else if (
+              birthEvent.status == "error" &&
+              birthEvent.message == "Forbidden"
+            ) {
+              userMessage(
+                "Du har inte behörighet att lägga till den här individen.",
+                "error"
+              );
+            } else {
+              userMessage(
+                // register_birth in data_access.py requires a litter size.
+                // Thus, the field must be mandatory in the form. Okay for föreningen?
+                "Något gick fel. Kolla så att du har lagt till mor, far, kullstorlek och födelsedatum.",
+                "error"
+              );
+            }
+          } else {
+            userMessage(
+              "Något gick fel på grund av tekniska problem. Kontakta en administrator.",
+              "error"
+            );
+          }
         } else {
-          userMessage("Individens nummer har inget giltigt format.", "warning");
+          userMessage(
+            "Första delen i individens nummer motsvarar ingen besättning.",
+            "warning"
+          );
```
I'm having a hard time keeping track of all the cases here, would it be possible to decrease the depth and have simple 
if cases that return instead?

HenrikeW committed:
remove bugs from error handling

pontus reviewed (approved):
Untested, but believe this will work.

HenrikeW committed:
Change logic of prepareIndividual

Invert the logic to present else statements first

HenrikeW committed:
minor logic fix

pontus committed:
Alternate example of how this can look

HenrikeW committed:
Merge pull request #329 from NBISweden/feature/breeding_practice_pontus

Alternate example of how this can look

HenrikeW committed:
some fixes for Pontus changes

HenrikeW committed:
some adjustments to Pontus changes

back to old error handling for/api/birth.
fix syntax for Map in createBreeding.
get rid of errors due type issues.

HenrikeW committed:
Merge pull request #331 from NBISweden/feature/breeding_practice

Feature/breeding practice

HenrikeW committed:
make change heading for selling date and herd

hopefully less confusion for the users

HenrikeW committed:
allow also matching birth date to find the right breeding
